packages/airnode-utilities/src/evm/gas-prices/gas-oracle.ts

```diff
@@ -0,0 +1,224 @@
+import { ethers } from 'ethers';
+import { go } from '@api3/promise-utils';
+import { config } from '@api3/airnode-validator';
+import { multiplyGasPrice, parsePriorityFee } from './gas-prices';
+import {
+  GAS_ORACLE_STRATEGY_MAX_TIMEOUT_MS,
+  GAS_ORACLE_STRATEGY_ATTEMPT_TIMEOUT_MS,
+  RANDOM_BACKOFF_MIN_MS,
+  RANDOM_BACKOFF_MAX_MS,
+} from '../../constants';
+import { logger, PendingLog, LogsData } from '../../logging';
+
+export const calculateTimeout = (startTime: number, totalTimeout: number) => totalTimeout - (Date.now() - startTime);
+
+export const getPercentile = (percentile: number, array: ethers.BigNumber[]) => {
+  if (!array.length) return;
+  array.sort((a, b) => (a.gt(b) ? 1 : -1));
+
+  const index = Math.ceil(array.length * (percentile / 100)) - 1;
+  return array[index];
+};
+
+// Check whether a value's change exceeds the maxDeviationMultipiler limit
+// and returns false if it does and true otherwise.
+export const checkMaxDeviationLimit = (
+  value: ethers.BigNumber,
+  referenceValue: ethers.BigNumber,
+  maxDeviationMultiplier: number
+) => {
+  // Handle maximum two decimals for maxDeviationMultiplier
+  const maxDeviationMultiplierBN = ethers.BigNumber.from(Math.round(maxDeviationMultiplier * 100));
+
+  return (
+    // Check that the current value is not higher than the maxDeviationMultiplier allows
+    referenceValue.gt(value.mul(ethers.BigNumber.from(100)).div(maxDeviationMultiplierBN)) &&
+    // Check that the current value is not lower than the maxDeviationMultiplier allows
+    referenceValue.lt(value.mul(maxDeviationMultiplierBN).div(ethers.BigNumber.from(100)))
+  );
+};
+
+// Returns the constant gas price
+export const fetchConstantGasPrice = (constantGasPriceStrategy: config.ConstantGasPriceStrategy) =>
+  parsePriorityFee(constantGasPriceStrategy.gasPrice);
+
+// Returns the provider gas price and applies the recommended multiplier
+export const fetchProviderRecommendedGasPrice = async (
+  provider: ethers.providers.StaticJsonRpcProvider,
+  gasOracleOptions: config.ProviderRecommendedGasPriceStrategy
+) => {
+  const { recommendedGasPriceMultiplier } = gasOracleOptions;
+
+  const startTime = Date.now();
+  const gasPrice = await go(() => provider.getGasPrice(), {
+    attemptTimeoutMs: GAS_ORACLE_STRATEGY_ATTEMPT_TIMEOUT_MS,
+    totalTimeoutMs: calculateTimeout(startTime, GAS_ORACLE_STRATEGY_MAX_TIMEOUT_MS),
```

Comment:
What is the `calculateTimeout` actually good for here? In Airseeker, we get the start time (current time) at the beginning of the processing and use it to calculate the timeout for all the chain calls down the line. Here, the `fetchProviderRecommendedGasPrice` function is called just once and you're getting the start time from within. `calculateTimeout` will just return pretty much the full `GAS_ORACLE_STRATEGY_MAX_TIMEOUT_MS` in this case.

You would have to get the start time before you start processing the various strategies and pass it to each of them that needs to make the chain calls so you will enforce one global timeout for gas price processing. If that's the intention.

Reply:
Hmm I see what you mean, thanks for clarifying. :sweat_smile: 

> If that's the intention.

Should that be the intention, or we could just let the higher level function `attemptGasOracleStrategy`  handle the timeouts and retry-delays, and the `provider.getGasPrice` and `provider.getBlockWithTransactions` **Go-Options** should only include retries?

Or do you have something else in mind?